contrib/groffer/groffer.man
.TH GROFFER @MAN1EXT@ "@MDATE@" "Groff Version @VERSION@"
.SH NAME
groffer \- display groff files and man\~pages on X and tty
.
.\" The .SH was moved to this place in order to appease `apropos'.
.
.
.\"*********************************************************************
.\"
.\" This man page doesn't use extended groff syntax!
.\" Formatting it with and without groff's option `-C' should always
.\" give the same result.
.\"
.\"*********************************************************************
.
.
.\" --------------------------------------------------------------------
.\" Legalese
.\" --------------------------------------------------------------------
.
.de co
Copyright \(co 2001-2014 Free Software Foundation, Inc.

This file is part of groffer, which is part of groff, a free software
project.

You can redistribute it and/or modify it under the terms of the GNU
General Public License version 2 as published by the Free Software
Foundation.

The license text is available in the internet at
.UR http://www.gnu.org/licenses/gpl-2.0.html
.UE .
..
.
.de au
This file was written by
.MT groff-bernd.warken-72@web.de
Bernd Warken
.ME .
..
.
.\" --------------------------------------------------------------------
.\" Characters
.\" --------------------------------------------------------------------
.
.\" Ellipsis ...
.ie t .ds EL \fS\N'188'\fP\"
.el .ds EL \&.\|.\|.\&\"
.\" called with \*(EL
.
.\" Bullet
.ie t .ds BU \(bu
.el .ds BU *
.\" used in `.IP \*(BU 2m' (former .Topic)
.
.
.\" --------------------------------------------------------------------
.SH "SYNOPSIS"
.\" --------------------------------------------------------------------
.
.SY groffer
.OP \-\-
.OP \%filespec \*(EL
.YS
.
.SY groffer
.OP mode-option \*(EL
.OP groff-options \*(EL
.OP man-options \*(EL
.OP X-options \*(EL
.OP \-\-
.OP \%filespec \*(EL
.YS
.
.SY groffer
.BR \-h " | " \-\-help
.YS
.
.SY groffer
.BR \-v " | " \-\-version
.YS
.
.
.\" --------------------------------------------------------------------
.SH DESCRIPTION
.\" --------------------------------------------------------------------
.
The
.B \%groffer
program is the easiest way to use
.BR \%groff (@MAN1EXT@).
It can display arbitrary documents written in the
.I \%groff
language, see
.BR \%groff (@MAN7EXT@),
or other
.I \%roff
languages, see
.BR \%roff (@MAN7EXT@),
that are compatible to the original
.I \%troff
language.
.
It finds and runs all necessary
.I groff
preprocessors, such as
.BR @g@chem .
.
.
.P
The
.B \%groffer
program also includes many of the features for finding and displaying
the \%\f(CRUnix\fP manual pages
.nh
.RI ( man\~pages ),
.hy
such that it can be used as a replacement for a
.BR \%man (1)
program.
.
Moreover, compressed files that can be handled by
.BR \%gzip (1)
or
.BR \%bzip2 (1)
are decompressed on-the-fly.
.
.
.P
The normal usage is quite simple by supplying a file name or name of a
.I \%man\~page
without further options.
.
But the option handling has many possibilities for creating special
behaviors.
.
This can be done either in configuration files, with the shell
environment variable
.SM
.BR \%$GROFFER_OPT ,
or on the command line.
.
.
.P
The output can be generated and viewed in several different ways
available for
.IR \%groff .
.
This includes the
.I \%groff
native \%\f(CRX\~Window\fP viewer
.BR \%gxditview (@MAN1EXT@),
each
.IR \%Postcript ,
.IR \%pdf ,
or
.I \%dvi
display program, a web browser by generating
.I \%html
in
.IR \%www\~mode ,
or several
.I \%text\~modes
in text terminals.
.
.
.P
Most of the options that must be named when running
.B \%groff
directly are determined automatically for
.BR \%groffer ,
due to the internal usage of the
.BR \%grog (@MAN1EXT@)
program.
.
But all parts can also be controlled manually by arguments.
.
.
.P
Several file names can be specified on the command line arguments.
.
They are transformed into a single document in the normal way of
.BR \%groff .
.
.
.P
Option handling is done in \f(CRGNU\fP style.
.
Options and file names can be mixed freely.
.
The option
.RB ` \-\- '
closes the option handling, all following arguments are treated as
file names.
.
Long options can be abbreviated in several ways.
.
.
.\" --------------------------------------------------------------------
.SH "OPTION OVERVIEW"
.\" --------------------------------------------------------------------
.
.TP
.I breaking options
.RS
.P
.SY
.OP \-h\~\fR|\fB\~\-\-help
.OP \-v\~\fR|\fB\~\-\-version
.YS
.RE
.
.
.TP
.I \%groffer mode options
.RS
.P
.SY
.OP \-\-auto
.OP \-\-default
.OP \-\-default\-modes mode1,mode2,\*(EL
.OP \-\-dvi
.OP \-\-groff
.OP \-\-html
.OP \-\-latin1
.OP \-\-mode display_mode
.OP \-\-pdf
.OP \-\-pdf2
.OP \-\-ps
.OP \-\-source
.OP \-\-text
.OP \-\-to\-stdout
.OP \-\-tty
.OP \-\-utf8
.OP \-\-viewer prog
.OP \-\-www
.OP \-\-x\~\fR|\fB\~\-\-X
.YS
.RE
.
.
.ig
Replace these options by --viewer
.OP \-\-dvi\-viewer prog
.OP \-\-html\-viewer prog
.OP \-\-pdf\-viewer prog
.OP \-\-ps\-viewer prog
.OP \-\-tty\-viewer prog
.OP \-\-www\-viewer prog
.OP \-\-x\-viewer\~\fR|\fB\~\-\-X\-viewer prog
..
.
.
.TP
.I options related to \%groff
.RS
.P
.SY
.OP \-T\~\fR|\fB\~\-\-device device
.OP \-Z\~\fR|\fB\~\-\-intermediate\-output\~\fR|\fB\~\-\-ditroff
.YS
.P
All further
.B \%groff
short options are accepted.
.RE
.
.
.TP
.I options for man\~pages
.RS
.P
.SY
.OP \-\-apropos
.OP \-\-apropos\-data
.OP \-\-apropos\-devel
.OP \-\-apropos\-progs
.OP \-\-man
.OP \-\-no\-man
.OP \-\-no\-special
.OP \-\-whatis
.YS
.RE
.
.
.TP
.I long options taken over from GNU man
.RS
.P
.SY
.OP \-\-all
.OP \-\-ascii
.OP \-\-ditroff
.OP \-\-extension suffix
.OP \-\-locale language
.OP \-\-local\-file
.OP \-\-location\~\fR|\fB\~\-\-where
.OP \-\-manpath dir1:dir2:\*(EL
.OP \-\-no\-location
.OP \-\-pager program
.OP \-\-sections sec1:sec2:\*(EL
.OP \-\-systems sys1,sys2,\*(EL
.OP \-\-troff\-device device
.YS
.P
Further long options of \f(CRGNU\fP
.B man
are accepted as well.
.RE
.
.
.TP
.I X Window Toolkit options
.RS
.P
.SY
.OP \-\-bd\~\fR|\fB\~\-\-bordercolor pixels
.OP \-\-bg\~\fR|\fB\~\-\-background color
.OP \-\-bw\~\fR|\fB\~\-\-borderwidth pixels
.OP \-\-display X-display
.OP \-\-fg\~\fR|\fB\~\-\-foreground color
.OP \-\-fn\~\fR|\fB\~\-\-ft\~\fR|\fB\~\-\-font font_name
.OP \-\-geometry size_pos
.OP \-\-resolution value
.OP \-\-rv
.OP \-\-title string
.OP \-\-xrm X\-resource
.YS
.RE
.
.
.TP
.I options for development
.RS
.P
.SY
.OP \-\-debug
.OP \-\-debug\-filenames
.OP \-\-debug\-grog
.OP \-\-debug\-keep
.OP \-\-debug\-params
.OP \-\-debug\-tmpdir
.OP \-\-do\-nothing
.OP \-\-print text
.OP \-V
.YS
.RE
.
.
.TP
.I \%filespec arguments
.RS
.P
The
.I \%filespec
parameters are all arguments that are neither an option nor an option
argument.
.
They usually mean a file name or a
.I man page
searching scheme.
.
.
.P
In the following, the term
.I section_extension
is used.
.
It means a word that consists of a
.I man section
that is optionally followed by an
.IR extension .
.
The name of a
.I man section
is a single character from
.BR \%[1\(en9on] ,
the
.I extension
is some word.
.
The
.I extension
is mostly lacking.
.
.
.P
No
.I \%filespec
parameters means standard input.
.
.
.TP 10m
.B \-
stands for standard input (can occur several times).
.
.
.TP
.I filename
the path name of an existing file.
.
.
.TP
.BI man: name ( section_extension )
.TQ
.BI man: name . section_extension
.TQ
.IB name ( section_extension )
.TQ
.IB name . section_extension
.TQ
.I "section_extension name"
search the \%man\~page
.I \%name
in the section with optional extension
.IR section_extension .
.
.
.TP
.BI man: name
\%man\~page in the lowest
.I \%man\~section
that has
.IR \%name .
.
.
.TP
.I name
if
.I \%name
is not an existing file search for the man\~page
.I \%name
in the lowest man\~section.
.
.RE
.
.
.\" --------------------------------------------------------------------
.SH "OPTION DETAILS"
.\" --------------------------------------------------------------------
.
The
.B \%groffer
program can usually be run with very few options.
.
But for special purposes, it supports many options.
.
These can be classified in 5 option classes.
.
.
.P
All short options of
.B \%groffer
are compatible with the short options of
.BR \%groff (@MAN1EXT@).
.
All long options of
.B \%groffer
are compatible with the long options of
.BR \%man (1).
.
.
.P
Arguments for long option names can be abbreviated in several ways.
.
First, the argument is checked whether it can be prolonged as is.
.
Furthermore, each minus sign
.B \-
is considered as a starting point for a new abbreviation.
.
This leads to a set of multiple abbreviations for a single argument.
.
For example,
.B \-\-de\-n\-f
can be used as an abbreviation for
.BR \-\-debug\-not\-func ,
but
.B \-\-de\-n
works as well.
.
If the abbreviation of the argument leads to several resulting options
an error is raised.
.
.
.P
These abbreviations are only allowed in the environment variable
.SM
.BR \%$GROFFER_OPT ,
but not in the configuration files.
.
In configuration, all long options must be exact.
.
.
.\" --------------------------------------------------------------------
.SS "groffer breaking Options"
.\" --------------------------------------------------------------------
.
As soon as one of these options is found on the command line it is
executed, printed to standard output, and the running
.B \%groffer
is terminated thereafter.
.
All other arguments are ignored.
.
.
.TP
.B \-h\~\fR|\fB\~\-\-help
Print help information with a short explanation of options to
standard output.
.
.
.TP
.B \-v\~\fR|\fB\~\-\-version
Print version information to standard output.
.
.
.\" --------------------------------------------------------------------
.SS "groffer Mode Options"
.\" --------------------------------------------------------------------
.
The display mode and the viewer programs are determined by these
options.
.
If none of these mode and viewer options is specified
.B \%groffer
tries to find a suitable display mode automatically.
.
The default modes are
.IR "mode pdf" ,
.IR "mode ps" ,
.IR "mode html" ,
.IR "mode x" ,
and
.I "mode dvi"
in \%\f(CRX\~Window\fP with different viewers and
.I mode tty
with device
.I utf8
under
.B less
on a terminal; other modes are tested if the programs for the main
default mode do not exist.
.
.
.P
In \%\f(CRX\~Window\fP, many programs create their own window when
called.
.
.B \%groffer
can run these viewers as an independent program in the background.
.
As this does not work in text mode on a terminal (tty) there must be a
way to know which viewers are \%\f(CRX\~Window\fP graphical
programs.
.
The
.B \%groffer
script has a small set of information on some viewer names.
.
If a viewer argument of the command\-line chooses an element that is
kept as \%\f(CRX\~Window\fP program in this list it is treated as a
viewer that can run in the background.
.
All other, unknown viewer calls are not run in the background.
.
.
.P
For each mode, you are free to choose whatever viewer you want.
.
That need not be some graphical viewer suitable for this mode.
.
There is a chance to view the output source; for example, the
combination of the options
.B \-\-mode=ps
and
.B \-\-viewer=less
shows the content of the
.I Postscript
output, the source code, with the pager
.BR less .
.
.
.TP
.B \-\-auto
Equivalent to
.BR \-\-mode=auto .
.
.
.TP
.B \-\-default
Reset all configuration from previously processed command line options
to the default values.
.
This is useful to wipe out all former options of the configuration, in
.SM
.BR \%$GROFFER_OPT ,
and restart option processing using only the rest of the command line.
.
.
.TP
.BI \-\-default\-modes \ mode1,mode2,\*(EL
Set the sequence of modes for
.I \%auto\~mode
to the comma separated list given in the argument.
.
See
.B \-\-mode
for details on modes.  Display in the default manner; actually, this
means to try the modes
.IR x ,
.IR ps ,
and
.I \%tty
in this sequence.
.
.
.
.TP
.B \-\-dvi
Equivalent to
.BR \-\-mode=\%dvi .
.TQ
.BI \-\-viewer \ prog
Choose a viewer program for
.IR \%dvi\~mode .
.
This can be a file name or a program to be searched in
.SM
.BR $PATH .
.
Known \%\f(CRX\~Window\fP
.I \%dvi
viewers include
.BR \%xdvi (1)
and
.BR \%dvilx (1).
.
In each case, arguments can be provided additionally.
.
.
.TP
.B \-\-groff
Equivalent to
.BR \-\-mode=groff .
.
.
.TP
.B \-\-html
Equivalent to
.BR \-\-mode=html .
.TQ
.B \-\-viewer
Choose a web browser program for viewing in
.IR \%html\~mode .
.
It can be the path name of an executable file or a program in
.BR $PATH .
.
In each case, arguments can be provided additionally.
.
.
.TP
.BI \-\-mode \ value
.
Set the display mode.
.
The following mode values are recognized:
.
.RS
.
.TP
.B auto
Select the automatic determination of the display mode.
.
The sequence of modes that are tried can be set with the
.B \-\-default\-modes
option.
.
Useful for restoring the
.I \%default\~mode
when a different mode was specified before.
.
.
.TP
.B dvi
Display formatted input in a
.I \%dvi
viewer program.
.
By default, the formatted input is displayed with the
.BR \%xdvi (1)
program.
.
.
.TP
.B groff
After the file determination, switch
.B \%groffer
to process the input like
.BR \%groff (@MAN1EXT@)
would do.
.
This disables the
.I \%groffer
viewing features.
.
.
.TP
.B html
Translate the input into html format and display the result in a web
browser program.
.
By default, the existence of a sequence of standard web browsers is
tested, starting with
.BR \%konqueror (1)
and
.BR \%mozilla (1).
The text html viewer is
.BR \%lynx (1).
.
.
.TP
.B pdf
Transform
.I roff input files
into a
.I PDF file
by using the
.B groff (@MAN1EXT@)
device
.BR -Tpdf .
.
This is the default
.B PDF
generator.
.
The generated
.I PDF file
is displayed with suitable viewer programs, such as
.BR okular (1).
.
.
.TP
.B pdf2
This is the traditional
.IR "pdf mode" .
.
Sometimes this mode produces more correct output than the default
.BR "PDF mode" .
.
By default, the input is formatted by
.B \%groff
using the Postscript device, then it is transformed into the PDF file
format using
.BR \%gs (1),
or
.BR ps2pdf (1).
.
If that's not possible, the
.I Postscript mode (ps)
is used instead.
.
Finally it is displayed using different viewer programs.
.
.
.TP
.B ps
Display formatted input in a Postscript viewer program.
.
By default, the formatted input is displayed in one of many viewer
programs.
.
.
.TP
.B text
Format in a
.I \%groff\~text\~mode
and write the result to standard output without a pager or viewer
program.
.
The text device,
.I \%latin1
by default, can be chosen with option
.BR \-T .
.
.
.TP
.B tty
Format in a
.I \%groff\~text\~mode
and write the result to standard output using a text pager program,
even when in \%\f(CRX\~Window\fP.
.
.
.TP
.B www
Equivalent to
.BR \-\-mode=html .
.
.
.TP
.B x
Display the formatted input in a native
.I roff
viewer.
.
By default, the formatted input is displayed with the
.BR \%gxditview (@MAN1EXT@)
program being distributed together with
.BR \%groff .
But the standard \%\f(CRX\~Window\fP tool
.BR \%xditview (1)
can also be chosen with the option
.BR \-\-viewer .
The default resolution is
.BR 75dpi ,
but
.B 100dpi
are also possible.
.
The default
.I groff
device
for the resolution of
.B 75dpi
is
.BR X75\-12 ,
for
.B 100dpi
it is
.BR X100 .
.
The corresponding
.I "groff intermediate output"
for the actual device is generated and the result is displayed.
.
For a resolution of
.BR 100dpi ,
the default width of the geometry of the display program is chosen to
.BR 850dpi .
.
.
.TP
.B X
Equivalent to
.BR \-\-mode=x .
.
.
.P
The following modes do not use the
.I \%groffer
viewing features.
.
They are only interesting for advanced applications.
.
.
.TP
.B groff
Generate device output with plain
.I \%groff
without using the special viewing features of
.IR \%groffer .
If no device was specified by option
.B \-T
the
.I \%groff
default
.B \%ps
is assumed.
.
.
.TP
.B source
Output the roff source code of the input files without further
processing.
.
.
.RE
.
.
.TP
.B \-\-pdf
Equivalent to
.BR \-\-mode=pdf .
.TQ
.B \-\-pdf2
Equivalent to
.BR \-\-mode=pdf2 .
.TQ
.BI \-\-viewer \ prog
Choose a viewer program for
.IR \%pdf\~mode .
.
This can be a file name or a program to be searched in
.SM
.BR $PATH ;
arguments can be provided additionally.
.
.
.TP
.B \-\-ps
Equivalent to
.BR \-\-mode=ps .
.TQ
.BI \-\-viewer \ prog
Choose a viewer program for
.IR \%ps\~mode .
.
This can be a file name or a program to be searched in
.SM
.BR $PATH .
.
Common Postscript viewers include
.BR \%okular (1),
.BR \%evince (1),
.BR \%gv (1),
.BR \%ghostview (1),
and
.BR \%gs (1),
.
In each case, arguments can be provided additionally.
.
.
.TP
.B \-\-source
Equivalent to
.BR \-\-mode=source .
.
.
.TP
.B \-\-text
Equivalent to
.BR \-\-mode=text .
.
.
.TP
.B \-\-to\-stdout
The file for the chosen mode is generated and its content is printed
to standard output.
.
It will not be displayed in graphical mode.
.
.
.TP
.B \-\-tty
Equivalent to
.BR \-\-mode=tty .
.TQ
.BI \-\-viewer \ prog
Choose a text pager for mode
.IR tty .
The standard pager is
.BR less (1).
This option is equivalent to
.I man
option
.BR \-\-pager=\,\fIprog\fP .
The option argument can be a file name or a program to be searched in
.SM
.BR $PATH ;
arguments can be provided additionally.
.
.
.TP
.B \-\-www
Equivalent to
.BR \-\-mode=html .
.TQ
.B \-\-viewer
.IR prog .
.
.
.TP
.B \-\-X\~\fR|\fB\~\-\-x
Equivalent to
.BR \-\-mode=x .
.TQ
.BI \-\-viewer " prog"
Choose a viewer program for
.IR \%x\~mode .
Suitable viewer programs are
.BR \%gxditview (@MAN1EXT@)
which is the default and
.BR \%xditview (1).
The argument can be any executable file or a program in
.SM
.BR $PATH ;
arguments can be provided additionally.
.
.
.TP
.B \-\-
Signals the end of option processing; all remaining arguments are
interpreted as
.I \%filespec
parameters.
.
.
.P
Besides these,
.B \%groffer
accepts all short options that are valid for the
.BR \%groff (@MAN1EXT@)
program.
.
All
.RB \%non- groffer
options are sent unmodified via
.B \%grog
to
.BR \%groff .
.
So postprocessors, macro packages, compatibility with
.I classical
.IR \%troff ,
and much more can be manually specified.
.
.
.\" --------------------------------------------------------------------
.SS "Options related to groff"
.\" --------------------------------------------------------------------
.
All short options of
.B \%groffer
are compatible with the short options of
.BR \%groff (@MAN1EXT@).
.
The following of
.B \%groff
options have either an additional special meaning within
.B \%groffer
or make sense for normal usage.
.
.
.P
Because of the special outputting behavior of the
.B \%groff
option
.B \-Z
.B \%groffer
was designed to be switched into
.IR \%groff\~mode ;
the
.I \%groffer
viewing features are disabled there.
.
The other
.B \%groff
options do not switch the mode, but allow to customize the formatting
process.
.
.
.TP
.B \-\-a
This generates an ascii approximation of output in the
.IR \%text\~modes .
.
That could be important when the text pager has problems with control
sequences in
.IR "tty mode" .
.
.
.TP
.BI \-\-m \ file
Add
.I \%file
as a
.I \%groff
macro file.
.
This is useful in case it cannot be recognized automatically.
.
.
.TP
.BI \-\-P \ opt_or_arg
Send the argument
.I \%opt_or_arg
as an option or option argument to the actual
.B \%groff
postprocessor.
.
.
.TP
.B \-\-T \fIdevname\fR\~\fR|\fB\~\-\-device \fIdevname\fR
.
This option determines
.BR \%groff 's
output device.
.
The most important devices are the text output devices for referring
to the different character sets, such as
.BR \%ascii ,
.BR \%utf8 ,
.BR \%latin1 ,
.BR \%utf8 ,
and others.
.
Each of these arguments switches
.B \%groffer
into a
.I \%text\~mode
using this device, to
.I \%mode\~tty
if the actual mode is not a
.IR \%text\~mode .
.
The following
.I \%devname
arguments are mapped to the corresponding
.B \%groffer
.B \-\-mode=\,\fIdevname\fR
option:
.BR \%dvi ,
.BR \%html ,
and
.BR \%ps .
All
.B \%X*
arguments are mapped to
.IR \%mode\~x .
Each other
.I \%devname
argument switches to
.I \%mode\~groff
using this device.
.
.
.TP
.B \-\-X
is equivalent to
.BR "groff \-X" .
It displays the
.I groff intermediate output
with
.BR gxditview .
As the quality is relatively bad this option is deprecated; use
.B \-\-X
instead because the
.I \%x\~mode
uses an
.IR X *
device for a better display.
.
.
.TP
.B \-Z\~\fR|\fB\~\-\-intermediate-output\~\fR|\fB\~\-\-ditroff
Switch into
.I \%groff\~mode
and format the input with the
.I \%groff intermediate output
without postprocessing; see
.BR \%groff_out (@MAN5EXT@).
This is equivalent to option
.B \-\-ditroff
of
.IR \%man ,
which can be used as well.
.
.
.P
All other
.B \%groff
options are supported by
.BR \%groffer ,
but they are just transparently transferred to
.B \%groff
without any intervention.
.
The options that are not explicitly handled by
.B \%groffer
are transparently passed to
.BR \%groff .
.
Therefore these transparent options are not documented here, but in
.BR \%groff (@MAN1EXT@).
Due to the automatism in
.BR \%groffer ,
none of these
.B \%groff
options should be needed, except for advanced usage.
.
.
.\" --------------------------------------------------------------------
.SS "Options for man\~pages"
.\" --------------------------------------------------------------------
.
.TP
.B \-\-apropos
Start the
.BR \%apropos (1)
command or facility of
.BR \%man (1)
for searching the
.I \%filespec
arguments within all
.I \%man\~page
descriptions.
.
Each
.I \%filespec
argument is taken for search as it is;
.I section
specific parts are not handled, such that
.B 7 groff
searches for the two arguments
.B 7
and
.BR groff ,
with a large result; for the
.I \%filespec
.B groff.7
nothing will be found.
.
The
.I language
locale is handled only when the called programs do support this; the
GNU
.B apropos
and
.B man \-k
do not.
.
The display differs from the
.B \%apropos
program by the following concepts:
.RS
.IP \*(BU 2m
Construct a
.I \%groff
frame similar to a
.I \%man\~page
to the output of
.BR \%apropos ,
.IP \*(BU 2m
each
.I \%filespec
argument is searched on its own.
.IP \*(BU 2m
The restriction by
.B \-\-sections
is handled as well,
.IP \*(BU 2m
wildcard characters are allowed and handled without a further option.
.RE
.
.
.TP
.B \-\-apropos\-data
Show only the
.B \%apropos
descriptions for data documents, these are the
.BR \%man (7)
.IR sections\~4 ", " 5 ", and " 7 .
.
Direct
.I section
declarations are ignored, wildcards are accepted.
.
.
.TP
.B \-\-apropos\-devel
Show only the
.B \%apropos
descriptions for development documents, these are the
.BR man (7)
.IR sections\~2 ", " 3 ", and " 9 .
.
Direct
.I section
declarations are ignored, wildcards are accepted.
.
.
.TP
.B \-\-apropos\-progs
Show only the
.B \%apropos
descriptions for documents on programs, these are the
.BR \%man (7)
.IR sections\~1 ", " 6 ", and " 8 .
.
Direct
.I section
declarations are ignored, wildcards are accepted.
.
.
.TP
.B \-\-whatis
For each
.I \%filespec
argument search all
.I \%man\~pages
and display their description \(em or say that it is not a
.IR \%man\~page .
This is written from anew, so it differs from
.IR man 's
.B whatis
output by the following concepts
.RS
.IP \*(BU 2m
each retrieved file name is added,
.IP \*(BU 2m
local files are handled as well,
.IP \*(BU 2m
the \fIlanguage\fP and \fIsystem\fP locale is supported,
.IP \*(BU 2m
the display is framed by a
.I groff
output format similar to a
.IR \%man\~page ,
.IP \*(BU 2m
wildcard characters are allowed without a further option.
.RE
.
.
.P
The following options were added to
.B \%groffer
for choosing whether the file name arguments are interpreted as names
for local files or as a search pattern for
.IR \%man\~pages .
.
The default is looking up for local files.
.
.
.TP
.B \-\-man
Check the non-option command line arguments
.nh
.RI ( filespecs )
.hy
first on being
.IR \%man\~pages ,
then whether they represent an existing file.
.
By default, a
.I \%filespec
is first tested whether it is an existing file.
.
.
.TP
.B \-\-no-man\~\fR|\fB\~\-\-local-file
Do not check for
.IR \%man\~pages .
.
.B \-\-local-file
is the corresponding
.B man
option.
.
.
.TP
.B \-\-no-special
Disable former calls of
.BR \-\-all ,
.BR \-\-apropos* ,
and
.BR \-\-whatis .
.
.
.\" --------------------------------------------------------------------
.SS "Long options taken over from GNU man"
.\" --------------------------------------------------------------------
.
The long options of
.B \%groffer
were synchronized with the long options of \f(CRGNU\fP
.BR man .
.
All long options of \f(CRGNU\fP
.B man
are recognized, but not all of these options are important to
.BR \%groffer ,
so most of them are just ignored.
.
These ignored
.B man
options are
.BR \-\-catman ,
.BR \-\-troff ,
and
.BR \-\-update .
.
.
.P
In the following, the
.B man

contrib/groffer/groffer.pl
#! /usr/bin/env perl

# groffer - display groff files

# Source file position: <groff-source>/contrib/groffer/groffer.pl
# Installed position: <prefix>/bin/groffer

# Copyright (C) 2006-2014
#   Free Software Foundation, Inc.

# Written by Bernd Warken <groff-bernd.warken-72@web.de>.

# This file is part of `groffer', which is part of `groff'.

# `groff' is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.

# `groff' is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program. If not, see
# <http://www.gnu.org/licenses/gpl-2.0.html>.

########################################################################

use strict;
use warnings;
#use diagnostics;

# temporary dir and files
use File::Temp qw/ tempfile tempdir /;

# needed for temporary dir
use File::Spec;

# for `copy' and `move'
use File::Copy;

# for fileparse, dirname and basename
use File::Basename;

# current working directory
use Cwd;

# $Bin is the directory where this script is located
use FindBin;


########################################################################
# system variables and exported variables
########################################################################

require 5.6.1;

our $Dev_Null;
our $Umask;
our @Path;
our $Start_Dir;

our $tmpdir = '';
our ($fh_cat, $tmp_cat);
our ($fh_stdin, $tmp_stdin);

our @Addopts_Groff;
our %Debug;
our %Opt;

our $Has_Compression;
our $Has_bzip;

our $Output_File_Name;

our $Apropos_Prog;
our $Filespec_Arg;
our $Filespec_Is_Man;
our $Macro_Pkg;
our $Manspec;
our $No_Filespecs;
our $Special_Filespec;
our $Special_Setup;

our %Man;

BEGIN {
  $Dev_Null = File::Spec->devnull();

  $Umask = umask 077;

  $Start_Dir = getcwd;

  # flush after each print or write command
  $| = 1;
}


########################################################################
# read-only variables with double-@ construct
########################################################################

our $File_split_env_sh;
our $File_version_sh;
our $Groff_Version;

my $before_make;		# script before run of `make'
{
  my $at = '@';
  $before_make = 1 if '@VERSION@' eq "${at}VERSION${at}";
}

my %at_at;
my $groffer_libdir;

if ($before_make) {
  my $groffer_source_dir = $FindBin::Bin;
  $at_at{'BINDIR'} = $groffer_source_dir;
  $at_at{'G'} = '';
  $at_at{'LIBDIR'} = '';
  $groffer_libdir = $groffer_source_dir;
  $File_version_sh = File::Spec->catfile($groffer_source_dir, 'version.sh');
  $Groff_Version = '';
} else {
  $Groff_Version = '@VERSION@';
  $at_at{'BINDIR'} = '@BINDIR@';
  $at_at{'G'} = '@g@';
  $at_at{'LIBDIR'} = '@libdir@';
  $groffer_libdir = '@groffer_dir@';
  $File_version_sh = File::Spec->catfile($groffer_libdir, 'version.sh');
}

die "$groffer_libdir is not an existing directory;"
  unless -d $groffer_libdir;

unshift(@INC, $groffer_libdir);

$File_split_env_sh = File::Spec->catfile($groffer_libdir, 'split_env.sh');
die "$File_split_env_sh does not exist;" unless -f "$File_split_env_sh";

require 'subs.pl';
require 'main_subs.pl';
require 'man.pl';

@Path = &path_uniq( File::Spec->path() );

if ( &where_is_prog('gzip') ) {
  $Has_Compression = 1;
  $Has_bzip = 1 if &where_is_prog('bzip2');
}


########################################################################
# modes, viewers, man sections, and defaults
########################################################################

# configuration files
our @Conf_Files = (File::Spec->catfile(File::Spec->rootdir(),
				      'etc', 'groff', 'groffer.conf'),
		  File::Spec->catfile("$ENV{'HOME'}", '.groff',
				      'groffer.conf')
		 );

our @Default_Modes = ('pdf', 'pdf2', 'html', 'ps', 'x', 'dvi', 'tty');
our $Default_Resolution = 100;
our $Default_tty_Device = 'utf8';

our @Macro_Packages = ('-man', '-mdoc', '-me', '-mm', '-mom', '-ms');

our %Viewer_tty = ('DVI' => [],
		  'HTML' => ['lynx', 'w3m'],
		  'PDF' => [],
		  'PS' => [],
		  'TTY' => ['less -r -R', 'more', 'pager'],
		  'X' => [],
		 );

our %Viewer_X = ('DVI' => ['kdvi', 'xdvi', 'dvilx'],
		'HTML' => ['konqueror', 'epiphany'. 'mozilla-firefox',
			   'firefox', 'mozilla', 'netscape', 'galeon',
			   'opera', 'amaya','arena', 'mosaic'],
		'PDF' => ['okular', 'kpdf', 'acroread', 'evince',
			  'xpdf -z 150', 'gpdf', 'xpdf', 'zathura'.
			  'epdfview', 'qpdfview', 'apvlv', 'qpdfview',
			  'kghostview --scale 1.45', 'gv', 'ggv'],
		'PS' => ['okular', 'evince', 'gv',
			 'gs', 'gs_x11', 'ghostscript', 'ghostview',
			 'kghostview --scale 1.45', 'ggv', 'kpdf'],
		'TTY' => ['xless'],
		'X' => ['gxditview', 'xditview'],
	      );

%Man = ('ALL' => 0,
	'AUTO_SEC' => ['1', '2', '3', '4', '5', '6', '7', '8', '9',
			  'n', 'o'],
	'ENABLE' => 1,
	   'EXT' => '',
	   'FORCE' => 0,
	   'IS_SETUP' => 0,
	   'MANSPEC' => {},
	   'LANG' => '',
	   'LANG2' => '',
	   'PATH' => [],
	   'SEC' => [],
	   'SEC_CHARS' => '',
	   'SYS' => [],
	  );
$Man{'AUTO_SEC_CHARS'} = join('', @{$Man{'AUTO_SEC'}});


########################################################################
# given options
########################################################################

our %Opts_Cmdline_Short;
our %Opts_Cmdline_Long;
our $Opts_Cmdline_Long_Str;
our %Opts_Cmdline_Double;
our %Opts_Groff_Short;

&main_set_options();


########################################################################
# $MANOPT
########################################################################

# handle environment variable $MANOPT
our @Manopt;

&main_parse_MANOPT();


########################################################################
# configuration files, $GROFFER_OPT, and command line, main_config_params()
########################################################################

our @Options;
our @Filespecs;
our @Starting_Conf;
our @Starting_ARGV = @ARGV;

&main_config_params();

if (0) {
  print STDERR "<$_>\n" foreach @ARGV;
}


########################################################################
# main_parse_params()
########################################################################

$Opt{'XRM'} = [];

our $i = 0;
our $n = $#Options;

&main_parse_params();


########################################################################
# main_set_mode()
########################################################################

our $Viewer_Background;
our $PDF_Did_Not_Work;
our $PDF_Has_gs;
our $PDF_Has_ps2pdf;
our %Display = ('MODE' => '',
	       'PROG' => '',
	       'ARGS' => ''
	      );

&main_set_mode();


########################################################################
# groffer temporary directory, main_temp()
########################################################################

&main_temp();


########################################################################
# tmp functions and compression
########################################################################

########################################################################
# main_do_fileargs() and related subs
########################################################################

our @REG_TITLE = ();

&main_do_fileargs();



########################################################################
# main_set_resources()
########################################################################

&main_set_resources();


########################################################################
# set resources
########################################################################

our $groggy;
our $modefile;
our $addopts;

&main_display();

&clean_up();


1;
########################################################################
### Emacs settings
# Local Variables:
# mode: CPerl
# End:
